### PR TITLE
Feat/enable custom dkim selector

### LINKED ISSUE #1304: Support custom DKIM selector
The current version of docker-mailserver always uses the value "mail" as DKIM selector. That works and (despite some confusion) is independent of host name. However, at least two users would like to make the DKIM selector configurable.

To do that:
- Add a DKIM_SELECTOR option with mail as the default value
- Modify the setup script to use the selector
- Modify the startup script to patch the configuration to use the selector
- Update the instructions
- Write some kind of test that proves it works

### PR DESCRIPTION
**Pull request by astrocket**
# Description

Enable generate-dkim-config to accept selector as third parameter.

```bash
./setup.sh config dkim <key-size> <domain.tld>[,<domain2.tld>] <selector>
```

which will apply to `private key generation`, `key names`, `KeyTable`, `SigningTable`

Fixes #1304

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Improvement (non-breaking change that does improve existing functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (README.md or ENVIRONMENT.md or the Wiki)
- [x] If necessary I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by astrocket** [comment]
Anyone has documentation for running test locally?

**Comment by aendeavor** [comment]
There's this thing, you know, called **_CONTRIBUTING.md_**. I assume you've _not_ read it...

---

Your addition is very welcome. But the way you've done it cannot work. You're using the selector as `$3` but the domains themself are actually optional. This way arguments will mix up. This script is not perfect, I mean before your change.

@wernerfred I thought about another way of parsing the arguments. I will explain in a few minutes in this thread.

**Comment by aendeavor** [comment]
From my PoV, the script should work like this:

``` MAN
SYNOPSIS
     ./setup.sh dkim [OPTIONS]

DESCRIPTION
    Configures DKIM keys. OPTIONS can be used to configure a more complex setup. LDAP setups require these options.

OPTIONS
  Generic Program Information
    help        Print the usage information.

  Configuration adjustments
    size         Set the size of the keys to be generated. Possible are 1024, 2024 (default) and 4096.
    selector   Set a manual selector (default is "mail") for the key.
    domains  Provide the domains for which keys are to be generated.

```

This adheres to a man-page style. I will provide a PR. This is a breaking change, but with the upcoming merge of #1802 and release `9.0.0`, this is fine.

**Comment by wernerfred** [comment]
I really like your approach. When I "hotfixed" the script I had no time to do a complete refactor. Your change will require many changes to tests too. Hit me up if you need support. If you use a branch and not a fork I can contribute too (maybe then it is also time to get rid of this ugly `wc -l` tests).

Regarding `9.0.0` we should definitely wait until all changes are merged. We now have a lot of PRs with breaking changes upfront.

**Comment by astrocket** [comment]
@aendeavor Thanks a lot :) I hope I can follow up.

**Comment by aendeavor** [comment]
> I really like your approach. When I "hotfixed" the script I had no time to do a complete refactor. Your change will require many changes to tests too. Hit me up if you need support. If you use a branch and not a fork I can contribute too (maybe then it is also time to get rid of this ugly `wc -l` tests).

I will open a PR  with on a branch and assign you too :D

> Regarding `9.0.0` we should definitely wait until all changes are merged. We now have a lot of PRs with breaking changes upfront.

Most definitely.

---

> @aendeavor Thanks a lot :) I hope I can follow up.

@astrocket We will get back to you once we're finished. Until then, I will mark this as a draft.

**Comment by aendeavor** [comment]
@wernerfred There is the new `origin/open-dkim` branch. I've refactored the whole script, and outsourced tests into their own script. #1812 is open :)

**Comment by aendeavor** [comment]
@astrocket #1812 was merged. You can go ahead and rebase on `master` (should require a force push).

Please note: Your script adjustments were fine. But the issue #1304 features a TODO list which is longer:

> To do that:
> 
> * Add a DKIM_SELECTOR option with mail as the default value
> * Modify the setup script to use the selector                                                 [DONE by me already]
> * Modify the startup script to patch the configuration to use the selector
> * Update the instructions
> * Write some kind of test that proves it works

**Comment by aendeavor** [comment]
Just for our (the maintainers) information: How long do you think it will take you to make this PR ready for merge @astrocket ?

**Comment by wernerfred** [comment]
I rebased the pr branch as the modified files were _all_ splitted in new files in the meantime. 
@astrocket pls have a look and @all pls review

File: test/open_dkim.bats
Comment (wernerfred) [inline test/open_dkim.bats:429]: These checks are not necessary imho as the latter ones check the same as the test above. It is enough to test if the provided selector is correctly used.

**Comment by wernerfred** [comment]
Any idea why Github only shows 2 changed files although the 4 commits edited 3 files?
image

**Comment by wernerfred** [comment]
Tests currently failing:

```
not ok 182 checking opendkim: generator creates keys, tables and TrustedHosts using manual provided selector name
# (from function `assert_output' in file test/test_helper/bats-assert/src/assert.bash, line 239,
#  in test file test/open_dkim.bats, line 421)
#   `assert_output 4' failed
# 
# -- output differs --
# expected : 4
# actual   : 1
# --
# 
```

**Comment by aendeavor** [comment]
LGTM 👍🏼

**Review by aendeavor — APPROVED** [review]

**Comment by wernerfred** [comment]
> Please note: Your script adjustments were fine. But the issue #1304 features a TODO list which is longer:
> 
> > To do that:
> > 
> > * Add a DKIM_SELECTOR option with mail as the default value
> > * Modify the setup script to use the selector                                                 [DONE by me already]
> > * Modify the startup script to patch the configuration to use the selector
> > * Update the instructions
> > * Write some kind of test that proves it works
> 

1, 2, 4 and 5 ist done. What needs to be done with 3?

**Comment by aendeavor** [comment]
I guess maybe

https://github.com/docker-mailserver/docker-mailserver/blob/9efa94ce6f5ff51e34cdf19e3497a69031e330b4/target/scripts/start-mailserver.sh#L1105-L1135

needs adjustments? I can't see any, but that doesn't mean they don't exist...

**Comment by wernerfred** [comment]
@astrocket did you test my changes again?

**Comment by wernerfred** [comment]
What do you think of changing this: 
https://github.com/docker-mailserver/docker-mailserver/blob/9efa94ce6f5ff51e34cdf19e3497a69031e330b4/target/scripts/start-mailserver.sh#L1105-L1135

to this:
```
function _setup_dkim
{
  _notify 'task' 'Setting up DKIM'

  mkdir -p /etc/opendkim

  # Check if any keys are available
  if [[ -e "/tmp/docker-mailserver/opendkim/KeyTable" ]]
  then
    cp -a /tmp/docker-mailserver/opendkim/* /etc/opendkim/

    _notify 'inf' "DKIM keys added for: $(ls -C /etc/opendkim/keys/)"
    _notify 'inf' "Changing permissions on /etc/opendkim"

    chown -R opendkim:opendkim /etc/opendkim/
    chmod -R 0700 /etc/opendkim/keys/ 
  else
    _notify 'warn' "No DKIM key provided. Check the documentation on how to get your keys."
  fi

  # setup nameservers paramater from /etc/resolv.conf if not defined
  if ! grep '^Nameservers' /etc/opendkim.conf
  then
    echo "Nameservers $(grep '^nameserver' /etc/resolv.conf | awk -F " " '{print $2}' | paste -sd ',' -)" >> /etc/opendkim.conf

    _notify 'inf' "Nameservers added to /etc/opendkim.conf"
  fi
}
```

Imho no need to:
1. `touch /etc/opendkim/SigningTable` as `SigningTable` gets copied later
2. `[[ ! -f ${KEYTABLE_FILE} ]] && touch "${KEYTABLE_FILE}" ` in case of no provided keys

**Comment by aendeavor** [comment]
If these changes work, go ahead :D

**Pull request merged**
